torch/csrc/profiler/unwind/unwind.cpp

```diff
@@ -0,0 +1,371 @@
+        refreshLibraries();
+        last_version_ = current_version;
+      }
+      auto& r = searchFor(addr);
+      if (addr >= r.last_addr()) {
+        throw std::runtime_error("addr not in range of known libraries");
+      }
+      return r;
+    }
+    return r;
+  }
+  torch::unwind::Stats stats() {
+    return stats_;
+  }
+
+ private:
+  const LibraryInfo& searchFor(uint64_t addr) {
+    uint64_t low = 0;
+    uint64_t high = all_libraries_.size();
+    while (low + 1 != high) {
+      auto mid = (low + high) / 2;
+      if (addr < all_libraries_.at(mid).load_bias()) {
+        high = mid;
+      } else {
+        low = mid;
+      }
+    }
+    return all_libraries_.at(low);
+  }
+
+  // sorted by load_bias
+  std::vector<LibraryInfo> all_libraries_;
+  ska::flat_hash_map<uint64_t, Unwinder> ip_cache_;
+
+  torch::unwind::Stats stats_;
+
+  // to keep track of whether we need to refresh this info
+  Version last_version_;
+};
+
+static UnwindCache unwind_cache;
+static std::shared_timed_mutex cache_mutex_;
+
+extern "C" void unwind_c(std::vector<void*>* result, int64_t rsp, int64_t rbp) {
+  std::shared_lock lock(cache_mutex_);
+  UnwindState state;
+  state.rip = *(int64_t*)(rsp);
+  state.rsp = rsp + 8;
+  state.rbp = rbp;
+  unwind_cache.checkRefresh(lock);
+  while (true) { // unwind for _start sets rip as being undefined
+    result->push_back((void*)state.rip);
+    const Unwinder& uw = unwind_cache.unwinderFor(state.rip, lock);
+    if (uw.terminator()) {
+      if (uw.isUnknown()) {
+        result->push_back(nullptr);
+      }
+      break;
+    }
+    state = uw.run(state);
+  }
+}
+
+extern "C" void unwind_entry(std::vector<void*>* result);
+
+__asm__(
+    ".global unwind_entry\n"
+    "unwind_entry:\n"
+    "mov %rsp, %rsi;\n"
+    "mov %rbp, %rdx;\n"
+    "jmp unwind_c;\n");
+
+namespace torch {
+namespace unwind {
+std::vector<void*> unwind() {
+  std::vector<void*> frames;
+  unwind_entry(&frames);
+  return frames;
+}
+
+// similar libraries: addr2line, llvm-symbolizer, gimli
+std::vector<Frame> symbolize(const std::vector<void*>& frames) {
+  // we need to make sure we don't write more than 64k bytes to
+  // a pipe before reading the results. Otherwise the buffer may
+  // get filled and block before we read the results.
+  // Each line is < 32 characters,
+  // so this limits us to < 32k bytes before we read rules.
+  constexpr int BLOCK = 1024;
+  struct Entry {
+    const LibraryInfo* lib;
+    std::unique_ptr<Communicate> comm;
+    std::vector<void*> queried;
+    size_t completed = 0;
+  };
+  std::vector<Entry> entries;
+  auto get_or_create = [&](const LibraryInfo& info) -> Entry& {
+    for (auto& e : entries) {
+      if (e.lib->load_bias() == info.load_bias()) {
+        return e;
+      }
+    }
+    const char* args[] = {
+        "addr2line", "-C", "-f", "-e", info.name().c_str(), nullptr};
+    entries.emplace_back(
+        Entry{&info, std::make_unique<Communicate>("addr2line", args), {}});
```

Comment:
I have a companion patch that uses the internal symbolizer code when this runs in fbcode but for OSS this seems like the most robust way to do it since it is so widely installed.